packages/elements-material-ui/src/Table/Table.tsx

```diff
@@ -0,0 +1,330 @@
+    (value: boolean) => {
+      if (props.hasOwnProperty('selectedRowIds')) {
+        const selectedIds = props.data.reduce((p, n: any) => ({ ...p, [n[props.rowKey]]: true }), {});
+        props.onRowSelected?.(value ? selectedIds : {});
+      } else {
+        toggleAllRowsSelected(value);
+      }
+    },
+    [props.onRowSelected]
+  );
+
+  const onRowSelected = useCallback(
+    (row: any, value: boolean) => {
+      const id = row[props.rowKey];
+      if (props.hasOwnProperty('selectedRowIds')) {
+        const newSelectedRows: any = { ...props.selectedRowIds };
+        if (value) {
+          newSelectedRows[id] = true;
+        } else {
+          delete newSelectedRows[id];
+        }
+        props.onRowSelected?.(newSelectedRows);
+      } else {
+        toggleRowSelected(id, value);
+      }
+    },
+    [props.onRowSelected]
+  );
+
+  useEffect(() => {
+    !props.hasOwnProperty('sortBy') && props.onSortChange?.(tableState.sortBy);
+  }, [props.sortBy, tableState.sortBy]);
+
+  useEffect(() => {
+    !props.hasOwnProperty('filters') && props.onFilterChange?.(tableState.filters);
+  }, [props.filters, tableState.filters]);
+
+  useEffect(() => {
+    tableRef.current?.querySelector('.fe-table__tbody')?.scroll?.({ top: 0, left: 0, behavior: 'smooth' });
+    props.onPageChange?.(tableState.pageSize, tableState.pageIndex);
+  }, [tableState.pageIndex]);
+
+  useEffect(() => {
+    !props.hasOwnProperty('selectedRowIds') && props.onRowSelected?.(tableState.selectedRowIds as any);
+  }, [tableState.selectedRowIds]);
+
+  const onPageChangeHandler = (page: number) => {
+    if (page > tableState.pageIndex) {
+      nextPage();
+    } else {
+      previousPage();
+    }
+  };
+
+  return (
+    <Paper className={classes.paper}>
+      <MaUTable className={classes.table} ref={tableRef} {...getTableProps()}>
+        <TableHead
+          headerGroups={headerGroups}
+          onSortChange={onSortChange}
+          onFilterChange={onFilterChange}
+          toggleAllRowsSelected={onToggleAllRowsSelected}
+          isAllRowsSelected={isAllRowsSelected}
+          selectedFlatRows={selectedFlatRows}
+        />
+        <TableBody
+          getTableBodyProps={getTableBodyProps}
+          prepareRow={prepareRow}
+          rows={(props.pagination ? page : rows) as (Row<T> & UseExpandedRowProps<T>)[]}
+          renderExpandedComponent={props.renderExpandedComponent}
+        />
+      </MaUTable>
+      {props.pagination === 'pages' && (
+        <TablePagination
+          className={classes.footer}
+          rowsPerPageOptions={[]}
+          component='div'
+          count={rows.length}
+          rowsPerPage={tableState.pageSize}
+          page={tableState.pageIndex}
+          onChangePage={(e, page) => onPageChangeHandler(page)}
+          ActionsComponent={(props) => (
+            <TablePaginationActions {...props} gotoPage={gotoPage} pageOptions={pageOptions} />
+          )}
+        />
+      )}
+    </Paper>
+  );
+};
```

Comment:
should be refactored, too big for one file

Reply:
I will export some of the functionalities to helpers and utils